File: src/main/java/nl/oxod/oxai/OxAI.java
package nl.oxod.oxai;

import net.fabricmc.api.ClientModInitializer;
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents;
import net.fabricmc.fabric.api.client.message.v1.ClientReceiveMessageEvents;
import net.minecraft.client.Minecraft;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.util.ArrayList;
import java.util.List;

public class OxAI implements ClientModInitializer {
  public static final String MODID = "oxai";
  private static final Logger logger = LoggerFactory.getLogger(MODID);

  private OxAIConfig config;
  private OllamaClient ollamaClient;

  private final List<String> messageQueue = new ArrayList<>();
  private int tickCounter = 0;

  @Override
  public void onInitializeClient() {
    logger.info("Initializing 0xAI...");

    config = OxAIConfig.load();
    ollamaClient = new OllamaClient(config);

    ClientReceiveMessageEvents.CHAT.register((message, signedMessage, messageType, senderUuid, params) -> {
      String messageText = message.getString();
      logger.info(messageText);
      if (messageText.toLowerCase().contains(config.getTrigger().toLowerCase())) {
        handleAIRequest(messageText);
      }
    });

    ClientTickEvents.END_CLIENT_TICK.register(client -> {
      tickCounter++;
      if (tickCounter >= 20 && !messageQueue.isEmpty()) {
        tickCounter = 0;
        String msg = sanitizeChatMessage(messageQueue.remove(0));
        sendChatMessage(client, msg);
      }
    });

    logger.info("0xAI initialized successfully!");
  }

  private void handleAIRequest(String message) {
    if (!config.isEnabled()) {
      logger.info("OxAI is disabled in config");
      return;
    }

    String question = message.replaceAll("(?i)" + config.getTrigger(), "").trim();
    if (question.isEmpty()) {
      logger.info("No question found, message only contained trigger");
      return;
    }

    new Thread(() -> {
      try {
        logger.info("Sending request to Ollama: " + question);
        String response = ollamaClient.generateResponse(message);
        for (String chunk : splitMessage(response)) {
          messageQueue.add(chunk);
        }
      } catch (Exception e) {
        logger.error("Failed to get response from Ollama", e);
        if (config.isShowErrors()) {
          messageQueue.add(config.getResponsePrefix() + "Error: " + e.getMessage());
        }
      }
    }).start();
  }

  private List<String> splitMessage(String message) {
    List<String> chunks = new ArrayList<>();
    int maxLength = config.getMaxMessageLength();
    String prefix = config.getResponsePrefix();

    if (message.length() + prefix.length() <= maxLength) {
      chunks.add(prefix + message);
      return chunks;
    }

    String remaining = message;
    while (!remaining.isEmpty()) {
      int availableLength = maxLength - prefix.length();
      int splitIndex = remaining.lastIndexOf(' ', availableLength);
      if (splitIndex == -1 || splitIndex > availableLength) {
        splitIndex = Math.min(availableLength, remaining.length());
      }
      chunks.add(prefix + remaining.substring(0, splitIndex));
      remaining = remaining.substring(splitIndex).trim();
    }
    return chunks;
  }

  private void sendChatMessage(Minecraft client, String message) {
    if (client.player != null && client.player.connection != null) {
      client.player.connection.sendChat(message);
    }
  }

  private String sanitizeChatMessage(String msg) {
    StringBuilder sb = new StringBuilder();
    for (char c : msg.toCharArray()) {
      if (c >= 32 && c <= 126) {
        sb.append(c);
      }
    }
    return sb.toString();
  }

  public static Logger getLogger() {
    return logger;
  }
}

File: src/main/kotlin/dev/oxod/oxai/OxAI.kt
package dev.oxod.oxai

import net.fabricmc.api.ClientModInitializer
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents
import net.fabricmc.fabric.api.client.message.v1.ClientReceiveMessageEvents
import net.minecraft.client.Minecraft
import org.slf4j.LoggerFactory

object OxAI : ClientModInitializer {
  const val MODID = "oxai"
  val logger by lazy { LoggerFactory.getLogger(MODID) }

  lateinit var config: OxAIConfig
  lateinit var ollamaClient: OllamaClient

  private val messageQueue by lazy { mutableListOf<String>() }
  private var tickCounter = 0

  override fun onInitializeClient() {
    logger.info("Initializing 0xAI...")

    config = OxAIConfig.load()
    ollamaClient = OllamaClient(config)

    ClientReceiveMessageEvents.CHAT.register(ClientReceiveMessageEvents.Chat { message, _, _, _, _ ->
      val messageText = message.string
      logger.info(messageText)
      if (messageText.contains(config.trigger, ignoreCase = true)) {
        handleAIRequest(messageText)
      }
    })

    ClientTickEvents.END_CLIENT_TICK.register { client ->
      tickCounter++
      if (tickCounter >= 20 && messageQueue.isNotEmpty()) {
        tickCounter = 0
        val msg = sanitizeChatMessage(messageQueue.removeAt(0))
        sendChatMessage(client, msg)
      }
    }

    logger.info("0xAI initialized successfully!")
  }

  private fun handleAIRequest(message: String) {
    if (!config.enabled) {
      logger.info("OxAI is disabled in config")
      return
    }

    val question = message.replace(config.trigger, "", ignoreCase = true).trim()
    if (question.isEmpty()) {
      logger.info("No question found, message only contained trigger")
      return
    }

    Thread {
      try {
        logger.info("Sending request to Ollama: $question")
        val response = ollamaClient.generateResponse(message)
        splitMessage(response).forEach { chunk ->
          messageQueue.add(chunk)
        }
      } catch (e: Exception) {
        logger.error("Failed to get response from Ollama", e)
        if (config.showErrors) {
          messageQueue.add("${config.responsePrefix}Error: ${e.message}")
        }
      }
    }.start()
  }

  private fun splitMessage(message: String): List<String> {
    val maxLength = config.maxMessageLength
    val prefix = config.responsePrefix

    if (message.length + prefix.length <= maxLength) return listOf("$prefix$message")

    val chunks = mutableListOf<String>()
    var remaining = message
    while (remaining.isNotEmpty()) {
      val availableLength = maxLength - prefix.length
      val splitIndex = remaining.lastIndexOf(' ', availableLength).takeIf { it != -1 } ?: availableLength
      chunks.add("$prefix${remaining.substring(0, splitIndex)}")
      remaining = remaining.substring(splitIndex).trim()
    }
    return chunks
  }

  private fun sendChatMessage(client: Minecraft, message: String) {
    client.player?.connection?.sendChat(message)
  }

  private fun sanitizeChatMessage(msg: String): String {
    return msg.filter { it.code in 32..126 }
  }
}
